crypto/fipsmodule/ecdh/ecdh.c

```diff
@@ -77,33 +77,44 @@
 #include "../ec/internal.h"
 
 
-int ECDH_compute_key_fips(uint8_t *out, size_t out_len, const EC_POINT *pub_key,
-                          const EC_KEY *priv_key) {
-  // Lock state here to avoid underlying |SHA*| functions updating the service
-  // indicator state unintentionally.
-  FIPS_service_indicator_lock_state();
-  int ret = 0;
+int ECDH_compute_shared_secret(uint8_t *buf, size_t *buflen, const EC_POINT *pub_key,
+                               const EC_KEY *priv_key) {
   if (priv_key->priv_key == NULL) {
     OPENSSL_PUT_ERROR(ECDH, ECDH_R_NO_PRIVATE_VALUE);
-    goto end;
+    return 0;
   }
   const EC_SCALAR *const priv = &priv_key->priv_key->scalar;
   const EC_GROUP *const group = EC_KEY_get0_group(priv_key);
   if (EC_GROUP_cmp(group, pub_key->group, NULL) != 0) {
     OPENSSL_PUT_ERROR(EC, EC_R_INCOMPATIBLE_OBJECTS);
-    goto end;
+    return 0;
   }
 
   EC_RAW_POINT shared_point;
-  uint8_t buf[EC_MAX_BYTES];
-  size_t buflen;
   if (!ec_point_mul_scalar(group, &shared_point, &pub_key->raw, priv) ||
-      !ec_get_x_coordinate_as_bytes(group, buf, &buflen, sizeof(buf),
+      !ec_get_x_coordinate_as_bytes(group, buf, buflen, *buflen,
                                     &shared_point)) {
     OPENSSL_PUT_ERROR(ECDH, ECDH_R_POINT_ARITHMETIC_FAILURE);
+    return 0;
+  }
+
+  return 1;
+}
+
+int ECDH_compute_key_fips(uint8_t *out, size_t out_len, const EC_POINT *pub_key,
+                          const EC_KEY *priv_key) {
+  uint8_t buf[EC_MAX_BYTES];
+  size_t buflen = sizeof(buf);
+  int ret = 0;
+
+  if (!ECDH_compute_shared_secret(buf, &buflen, pub_key, priv_key)) {
     goto end;
   }
 
+  // Lock state here to avoid underlying |SHA*| functions updating the service
+  // indicator state unintentionally.
+  FIPS_service_indicator_lock_state();
```

Comment:
Why was this moved down? In theory could ECDH_compute_shared_secret trigger something?